### DISCUSSION TITLE
Utilizing OpenRewrite without using the build tool plugin/creating a .jar file

### DISCUSSION THREAD

**Discussion by seanpmyers** [openrewrite/rewrite#3079] [Q&A] [answered]
# Utilizing OpenRewrite without using the build tool plugin/creating a .jar file

## Questions

Hello, apologies for my ignorance but I was wondering if someone could quickly help me to answer the two following questions:

1. Is it possible to use **_ALL_** openrewrite recipes without using the build tool plugin (in other words by just importing the libraries via maven/gradle)?
2. Is it possible to use Java recipes without creating a `.jar` file for a project?

## Goal

My goal here is to be able to clone a repository and apply a recipe without building the project/`.jar` file.

I currently have a proof of concept console app where I am importing the maven dependencies and I am not using the build tool plugin.  
I followed the [example documentation](https://docs.openrewrite.org/running-recipes/running-rewrite-without-build-tool-plugins) for running openrewrite without the build tool plugin.  

## Current Issues

My proof of concept has two issue:

1. I don't want to have to build the target project's `.jar` file before applying a selected Java recipe.
2. When I tried importing the `yaml` recipe without the build tool plugin, I was having difficulty getting it to work (I was trying to directly access methods like [changeValue](https://github.com/openrewrite/rewrite/blob/main/rewrite-yaml/src/main/java/org/openrewrite/yaml/ChangeValue.java)). 

Any feedback would be appreciated. 

Thanks!

**Comment by timtebeek** [comment, marked as answer]
hi @seanpmyers ; Sounds like you're building a POC towards a tool that we already have internally, and make available to our customers: [the `moderne-cli`](https://docs.moderne.io/moderne-cli/cli-intro). This abstracts over project specifics such as Maven/Gradle/other, and allows users to easily ingest projects into their moderne instance, and quickly run recipes across a broad set of projects. Perhaps we can have you up and running a lot quicker without duplicating any effort with this existing CLI.

Could you tell us a bit more about the context of your POC, what types and number or projects you aim to run this on?
We hope to learn from your answers how we can improve our CLI, to make it fit your purpose as well.

Remember that the [documentation on running without build tool plugins](https://docs.openrewrite.org/running-recipes/running-rewrite-without-build-tool-plugins) starts with the following caveat; it's not a trivial undertaking, and could prove limiting if you want to apply this broadly.

> It's simple to run Rewrite from a regular Java main method outside of the build tool plugins. The hardest part is determining the classpath that is relevant for each set of Java files. In the case of Maven and Gradle projects, locating each source set and the classpath that applies to them is non-trivial enough that it's much easier to use the build tool plugins instead.
> 
> This tutorial is useful primarily for projects that have a simpler structure then, e.g. a project with a simple source directory and a lib directory with all the JARs for its classpath already downloaded.
